core/client/fs/src/main/java/alluxio/client/file/cache/DefaultPageMetaStore.java

```diff
@@ -20,40 +20,72 @@
 
 import com.codahale.metrics.Counter;
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
 
 /**
  * The default implementation of a metadata store for pages stored in cache. This implementation
- * is not thread safe and requires synchronizations on external callers.
+ * is not thread safe and requires synchronizations on external callers by acquiring the associated
+ * lock.
  */
+@NotThreadSafe
 public class DefaultPageMetaStore implements PageMetaStore {
   private static final Logger LOG = LoggerFactory.getLogger(DefaultPageMetaStore.class);
   /** A map from PageId to page info. */
   private final Map<PageId, PageInfo> mPageMap = new HashMap<>();
+  private final ImmutableList<PageStoreDir> mDirs;
   /** The number of logical bytes used. */
   private final AtomicLong mBytes = new AtomicLong(0);
 
+  protected final ReentrantReadWriteLock mLock;
+
   /**
    * Constructor of DefaultMetaStore.
+   *
+   * @param dirs storage directories
    */
-  public DefaultPageMetaStore() {
+  public DefaultPageMetaStore(List<PageStoreDir> dirs) {
+    this(dirs, new ReentrantReadWriteLock());
     //metrics for the num of pages stored in the cache
     MetricsSystem.registerGaugeIfAbsent(MetricKey.CLIENT_CACHE_PAGES.getName(),
         mPageMap::size);
   }
 
+  /**
+   * Creates a new instance with a shared lock.
+   *
+   * @param dirs storage directories
+   * @param lock the lock to associate with this meta store
+   */
+  public DefaultPageMetaStore(List<PageStoreDir> dirs, ReentrantReadWriteLock lock) {
+    mLock = lock;
+    mDirs = ImmutableList.copyOf(dirs);
```

Comment:
if we want a guarantee, we could using the ImmutableList when we populate the PageStoreDir list

Reply:
`PageStoreDir#createPageStoreDirs` indeed uses `ImmutableList`, and `ImmutableList.copyOf` actually avoids a copy when the list is already an `ImmutableList`.

The javadoc of `ImmutableList.copyOf` says:

> Despite the method name, this method attempts to avoid actually copying the data when it is safe to do so. The exact circumstances under which a copy will or will not be performed are undocumented and subject to change.